src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -78,18 +133,83 @@ etaConversionPass ast = do
+    , entryArgTypes   = map IR.varPatType (IR.funcDeclArgs newFuncDecl)
+    , entryReturnType = IR.funcDeclReturnType newFuncDecl
+    }
+  return newFuncDecl
+
+-- | Compute the new function declaration where all missing top-level 
+--   arguments have been added and all occurring functions are fully applied.  
+modifyTopLevel :: IR.FuncDecl -> IR.Expr -> [String] -> Converter IR.FuncDecl
+modifyTopLevel funcDecl rhs newArgIdents = do
+  -- Compute the function's new (uncurried) type. Assumes that funcDecl's return type is known.
+  (newArgTypes, returnType) <- splitFuncType
+    (IR.funcDeclQName funcDecl)
+    (map IR.toVarPat newArgIdents)
+    (fromJust $ IR.funcDeclReturnType funcDecl)
+  -- Compute the function's new arguments and add them to the argument list.
+  let newArgs =
+        zipWith (IR.VarPat NoSrcSpan) newArgIdents (map Just newArgTypes)
+  let vars' = IR.funcDeclArgs funcDecl ++ newArgs
+  -- Compute the new right-hand side.
+  rhs' <- etaConvertTopLevel newArgs rhs
+  -- Update the function declaration's attributes.
+  return funcDecl { IR.funcDeclArgs       = vars'
+                  , IR.funcDeclReturnType = Just returnType
+                  , IR.funcDeclRhs        = rhs'
+                  }
+
 
 -------------------------------------------------------------------------------
 -- Expressions                                                               --
 -------------------------------------------------------------------------------
 
--- | Applies η-conversions to the given expression and it's sub-expressions
+-- | Apply missing arguments to all top-level alternatives of an expression
+--   and perform further necessary eta-conversions afterwards.
+etaConvertTopLevel :: [IR.VarPat] -> IR.Expr -> Converter IR.Expr
+-- If there is more than one alternative, apply the conversion to 
+-- all alternatives. 
+etaConvertTopLevel argPats expr@(IR.If _ _ _ _ _) =
+  etaConvertAlternatives argPats expr
+etaConvertTopLevel argPats expr@(IR.Case _ _ _ _) =
+  etaConvertAlternatives argPats expr
+-- If there is only one alternative, apply it to the newly-added arguments, then 
+-- apply @etaConvertExpr@ to it to make so the expression and its sub-expressions 
+-- are fully applied.
+etaConvertTopLevel argPats expr = localEnv $ do
+  let argExprs = map IR.varPatToExpr argPats
+  -- apply expression to missing arguments and perform eta-conversion on the 
+  -- resulting expression 
+  etaConvertExpr $ IR.app NoSrcSpan expr argExprs
+
+-- | Calls @etaConvertTopLevel@ on all alternatives in an if or case expression. 
+etaConvertAlternatives :: [IR.VarPat] -> IR.Expr -> Converter IR.Expr
+etaConvertAlternatives argPats expr = do
+  -- The first child term of an if or case expression is the condition/the scrutinee
+  -- and should remain unchanged.
+  let (e : subterms) = childTerms expr
+  subterms' <- mapM (etaConvertTopLevel argPats) subterms
+  let Just expr' = replaceChildTerms expr (e : subterms')
```

Comment:
Maybe we should discuss, if we want to just `fromJust` directly or use the let-binding like here. Any arguments on your or @just95's side?